lib/spack/spack/ci.py

```diff
@@ -2147,39 +2141,61 @@ def process_command(name, commands, repro_dir):
     return exit_code
 
 
-def create_buildcache(**kwargs):
+def create_buildcache(input_spec: spack.spec.Spec, **kwargs):
     """Create the buildcache at the provided mirror(s).
 
     Arguments:
-       kwargs (dict): dictionary of arguments used to create the buildcache
+        input_spec: Installed spec to package and push
+        kwargs (dict): dictionary of arguments used to create the buildcache
 
     List of recognized keys:
-
-    * "env" (spack.environment.Environment): the active environment
     * "buildcache_mirror_url" (str or None): URL for the buildcache mirror
     * "pipeline_mirror_url" (str or None): URL for the pipeline mirror
     * "pr_pipeline" (bool): True if the CI job is for a PR
-    * "json_path" (str): path the the spec's JSON file
+
+    Returns:
+        List of objects indicating success or failure for each attempted
+        push.  Each object has the form:
+
+        .. code-block:: JSON
+
+           {
+              "success": True|False,
+              "url": mirror_url of attempted push
+           }
+
     """
-    env = kwargs.get("env")
     buildcache_mirror_url = kwargs.get("buildcache_mirror_url")
     pipeline_mirror_url = kwargs.get("pipeline_mirror_url")
     pr_pipeline = kwargs.get("pr_pipeline")
-    json_path = kwargs.get("json_path")
 
     sign_binaries = pr_pipeline is False and can_sign_binaries()
 
+    results = []
+
     # Create buildcache in either the main remote mirror, or in the
     # per-PR mirror, if this is a PR pipeline
     if buildcache_mirror_url:
-        push_mirror_contents(env, json_path, buildcache_mirror_url, sign_binaries)
+        results.append(
+            {
+                "success": push_mirror_contents(input_spec, buildcache_mirror_url, sign_binaries),
+                "url": buildcache_mirror_url,
+            }
+        )
 
     # Create another copy of that buildcache in the per-pipeline
     # temporary storage mirror (this is only done if either
     # artifacts buildcache is enabled or a temporary storage url
     # prefix is set)
     if pipeline_mirror_url:
-        push_mirror_contents(env, json_path, pipeline_mirror_url, sign_binaries)
+        results.append(
+            {
+                "success": push_mirror_contents(input_spec, pipeline_mirror_url, sign_binaries),
```

Comment:
Sorry, but why do you use dictionaries all over the place, whether it's return types or **kwargs. They use type info. Please stop flooding the Spack code base with them. Use named tuples or a small class instead.